internal/envoy/config/config.go

```diff
@@ -130,37 +133,79 @@ func (e *EnvoyConfiguration) ClusterExist(name string) bool {
 // For the simplicity right now we don't support endpoints assignments separately, i.e. one cluster - one endpoint, not multiple load balanced
 // Cluster with the same name will be overwritten
 func (e *EnvoyConfiguration) AddCluster(clusterName, upstreamServiceHost string, upstreamServicePort uint32) {
+	loadAssignment := createLoadAssignment(clusterName, upstreamServiceHost, upstreamServicePort)
+
+	e.clusters[clusterName] = &cluster.Cluster{
+		Name:                 clusterName,
+		ConnectTimeout:       &durationpb.Duration{Seconds: 5},
+		ClusterDiscoveryType: &cluster.Cluster_Type{Type: cluster.Cluster_LOGICAL_DNS},
+		LbPolicy:             cluster.Cluster_ROUND_ROBIN,
+		LoadAssignment:       loadAssignment,
+		DnsLookupFamily:      cluster.Cluster_V4_ONLY,
+	}
+}
+
+// AddClusterWithTLS - AddCluster with SNI or rather `AutoSni` enabled`.
+// Example SNI : "kubeshop-kusk-gateway-oauth2.eu.auth0.com" -> "eu.auth0.com"
+func (e *EnvoyConfiguration) AddClusterWithTLS(clusterName, upstreamServiceHost string, upstreamServicePort uint32) error {
+	loadAssignment := createLoadAssignment(clusterName, upstreamServiceHost, upstreamServicePort)
+
+	upstreamTlsContext := &envoy_extensions_transport_sockets_tls_v3.UpstreamTlsContext{}
+	anyUpstreamTlsContext, err := anypb.New(upstreamTlsContext)
+	if err != nil {
+		return fmt.Errorf("EnvoyConfiguration.AddClusterWithTLS: failed on `anypb.New(upstreamTlsContext)`, %w", err)
+	}
+
+	transportSocket := &core.TransportSocket{
+		Name: "envoy.transport_sockets.tls",
+		ConfigType: &core.TransportSocket_TypedConfig{
+			TypedConfig: anyUpstreamTlsContext,
+		},
+	}
+
+	e.clusters[clusterName] = &cluster.Cluster{
+		Name:                 clusterName,
+		ConnectTimeout:       &durationpb.Duration{Seconds: 5},
+		ClusterDiscoveryType: &cluster.Cluster_Type{Type: cluster.Cluster_LOGICAL_DNS},
+		LbPolicy:             cluster.Cluster_ROUND_ROBIN,
+		LoadAssignment:       loadAssignment,
+		DnsLookupFamily:      cluster.Cluster_V4_ONLY,
+		TransportSocket:      transportSocket,
+		UpstreamHttpProtocolOptions: &core.UpstreamHttpProtocolOptions{
+			AutoSni: true,
```

Comment:
What exactly does AutoSni do?

Reply:
TLS is required to reach the `token_endpoint` which is generally on HTTPS. So this `cluster` requires certain additional configuration parameters such as `AutoSni` or specifying the `sni` directly, which requires more code so I opted to just use the `AutoSni` feature. The value of `sni` would need to be calculated and if you read the the function documentation of:

```
// AddClusterWithTLS - AddCluster with SNI or rather `AutoSni` enabled`.
// Example SNI : "kubeshop-kusk-gateway-oauth2.eu.auth0.com" -> "eu.auth0.com"
(e *EnvoyConfiguration) AddClusterWithTLS(clusterName, upstreamServiceHost string, upstreamServicePort uint32) error
```

You have to deduce `"eu.auth0.com"` from `"kubeshop-kusk-gateway-oauth2.eu.auth0.com"`.

I've documented what `AutoSni` does, and a safety measure added a call to `cluster.ValidateAll()` in the `AddClusterWithTLS` function to the `cluster` that has `AutoSni` set to true.

Reply:
that's plenty